File: modules/user/user.mapper.go
package user

import (
	"github.com/yonsina94/go-user-manager/modules/commons/transformer"
	"github.com/yonsina94/go-user-manager/modules/user/entities"
)

// UserMapper es el transformador sin estado para la entidad User y su DTO.
type UserMapper struct{}

// NewUserMapper crea una nueva instancia de UserMapper que implementa la interfaz transformer.Transformer.
func NewUserMapper() transformer.Transformer[entities.User, UserDTO] {
	return &UserMapper{}
}

// ToDTO convierte una entidad User en su DTO correspondiente.
func (UserMapper) ToDTO(u entities.User) UserDTO {
	return ToUserDTO(u)
}

// ToEntity convierte un UserDTO en su entidad User correspondiente.
func (UserMapper) ToEntity(d UserDTO) entities.User {
	return entities.User{
		Name:     d.Name,
		LastName: d.LastName,
		Username: d.Username,
		Email:    d.Email,
		Role:     d.Role,
		Active:   d.Active,
	}
}

File: frontend/user-manager-frontend/src/mappers/user.mapper.ts
import type { User, UserDTO } from "../types/user";

export const mapUserDtoToUser = (dto: UserDTO): User => {
    return {
        id: dto.id,
        name: dto.name,
        lastname: dto.lastName,
        username: dto.username,
        email: dto.email,
        role: dto.role,
        active: dto.active,
        createAt: dto.createdAt ? new Date(dto.createdAt) : new Date(),
    };
};

export const mapUserDtosToUsers = (dtos: UserDTO[]): User[] => {
    return dtos.map(mapUserDtoToUser);
};

export const mapUserToUserDto = (user: User): UserDTO => {
    return {
        id: user.id,
        name: user.name,
        lastName: user.lastname,
        username: user.username,
        email: user.email,
        role: user.role,
        active: user.active,
        createdAt: user.createAt.toISOString(),
    };
};

export const mapUsersToDtos = (users: User[]): UserDTO[] => {
    return users.map(mapUserToUserDto);
};